### PR TITLE
feat: initial charm

### PR DESCRIPTION
**Pull request by yanksyoon**
Applicable spec: ISD-143

### Overview

- A charm that periodically builds ubuntu image.
- Supports relation to GitHub runner and provides the image ID on openstack for consumption.

### Rationale

- To support GitHub runner with ready-to-use images.

### Juju Events Changes

- On install: installs packages needed to build custom images.
- On config changed: builds new image according to the config and resets cron if necessary.
- On build-image action: juju action to manually trigger a new image build.
- On cron-trigger: custom internal hook to enable rebuild with cron jobs, propagating new images to existing relations.
- On image relation joined: provides any existing latest image to the relation if available.

### Module Changes

- builder: responsible for building images
- charm: main charm event handlers
- chroot: module for handling chroot environments
- cron: module for handling cron hooks
- image: observer module for handling image relation event handlers
- openstack_manager: module responsible for communicating with openstack
- state: the charm state
- utils: provides retry

### Library Changes

- Uses operator libs linux.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [ ] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

Due to some of the issues with ARM64 private-endpoint runners, the charm has been manually tested.
image
image
image

The command tests can be found on the PR of the application here: https://github.com/canonical/github-runner-image-builder/pull/4

### CODE REVIEW

File: .github/workflows/integration_test.yaml [hunk 1]
```diff
@@ -8,17 +8,11 @@ jobs:
...
-      zap-cmd-options: '-T 60 -z "-addoninstall jython" --hook "/zap/wrk/tests/zap/hook.py"'
-      zap-target: localhost
-      zap-target-port: 80
-      zap-rules-file-name: "zap_rules.tsv"
-      trivy-fs-enabled: true
-      trivy-image-config: "trivy.yaml"
+      juju-channel: 3.2/stable
+      pre-run-script: tests/integration/pre-run.sh
+      provider: microk8s
+      channel: 1.26-strict/stable
```
Comment (jdkandersson) [inline .github/workflows/integration_test.yaml:14]: I think there is a later version of microk8s available

File: .github/workflows/test.yaml
Comment (jdkandersson) [inline .github/workflows/test.yaml:1]: SHould we fix this in the template repo?

File: .dockerignore
Comment (jdkandersson) [inline .dockerignore:1]: Should we remove this in the template repo

File: .woke.yaml [hunk 2]
```diff
@@ -1,5 +1,7 @@
 ignore_files:
 # List of files to ignore
+  - lib/*
 
 rules:
-# List of words to ignore
+  # ignore blacklist, a term used by bandit
+  - name: blacklist
```
Comment (jdkandersson) [inline .woke.yaml:7]: Final new line?

File: charmcraft.yaml [hunk 3]
```diff
@@ -1,11 +1,83 @@
...
+  website: https://charmhub.io/github-runner-image-builder
 
 type: charm
 bases:
   - build-on:
-    - name: ubuntu
+    - name: "ubuntu"
       channel: "22.04"
+      architectures:
+      - amd64
```
Comment (jdkandersson) [inline charmcraft.yaml:34]: arm64?

File: charmcraft.yaml [hunk 4]
```diff
@@ -1,11 +1,83 @@
...
     run-on:
-    - name: ubuntu
+    - name: "ubuntu"
+      channel: "22.04"
+      architectures:
+      - amd64
```
Comment (jdkandersson) [inline charmcraft.yaml:39]: arm64?

File: charmcraft.yaml [hunk 5]
```diff
@@ -1,11 +1,83 @@
...
+      description: |
+        The base ubuntu OS image to use for the runners. Codename (e.g. "jammy") or version tag
+        (e.g. 22.04) is supported as input. Currently only supports LTS versions of jammy and
+        higher, i.e. jammy, noble.
+    build-interval:
+      type: int
+      default: 6
+      description: |
+        The interval in hours between each scheduled image builds. Setting this value to 0 means
+        image will not be rebuilt after the initial build.
```
Comment (jdkandersson) [inline charmcraft.yaml:65]: Why support 0?

File: pyproject.toml [hunk 6]
```diff
@@ -12,31 +21,37 @@ log_cli_level = "INFO"
...
 line-length = 99
-target-version = ["py38"]
+target-version = ["py310"]
+
+[tool.isort]
+line_length = 99
+profile = "black"
 
 # Linting tools configuration
-[tool.ruff]
```
Comment (jdkandersson) [inline pyproject.toml:18]: Perhaps we fix this in the template repo

File: src/charm.py [hunk 7]
```diff
@@ -1,103 +1,141 @@
 #!/usr/bin/env python3
-# Copyright 2023 Mariyan Dimitrov
```
Comment (jdkandersson) [inline src/charm.py:2]: Should we fix this in the template repo?

File: src/charm.py [hunk 8]
```diff
@@ -1,103 +1,141 @@
...
+import image
+import openstack_manager
+from state import CharmConfigInvalidError, CharmState
+
 logger = logging.getLogger(__name__)
 
-VALID_LOG_LEVELS = ["info", "debug", "warning", "error", "critical"]
 
+class GithubRunnerImageBuilderCharm(ops.CharmBase):
+    """Charm the service.
```
Comment (jdkandersson) [inline src/charm.py:23]: Should we update this?

File: src/charm.py [hunk 9]
```diff
@@ -1,103 +1,141 @@
...
+        self.image_observer = image.Observer(self)
+        self.framework.observe(self.on.install, self._on_install)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.build_image_action, self._on_build_image_action)
+        self.framework.observe(self.on.trigger, self._on_cron_trigger)
 
-    def _on_httpbin_pebble_ready(self, event: ops.PebbleReadyEvent):
-        """Define and start a workload using the Pebble API.
+    def _load_state(self) -> CharmState | None:
+        """Load the charm state if valid, set charm to blocked if otherwise.
```
Comment (jdkandersson) [inline src/charm.py:45]: Perhaps just otherwise?

File: src/charm.py [hunk 10]
```diff
@@ -1,103 +1,141 @@
...
+        if not state:
+            # Defer this event since on install should be re-triggered to setup dependencies for
+            # the charm. Since the charm goes into blocked state and the user reconfigures the
+            # state, config_changed event should be queued after the deferred on_install.
+            event.defer()
+            return
+
+        builder.configure_proxy(proxy=state.proxy_config)
+        builder.setup_builder()
+        self.unit.status = ops.WaitingStatus("Waiting for first image build.")
```
Comment (jdkandersson) [inline src/charm.py:75]: As discussed, this should probably be active since the workload is up and running and busy building the first image

Comment (jdkandersson) [reply src/charm.py:75]: Perhaps Waiting for first image could be the status message

File: src/charm.py [hunk 11]
```diff
@@ -1,103 +1,141 @@
...
+            base=state.image_config.base_image,
+            num_revisions=state.revision_history_limit,
+            src_path=image_path,
+        )
+        self.unit.status = ops.StatusBase.from_name(self.unit.status.name, "Updating image.")
+        with openstack_manager.OpenstackManager(cloud_config=state.cloud_config) as openstack:
+            image_id = openstack.upload_image(config=upload_config)
+
+        self.image_observer.update_relation_data(image_id=image_id)
+        return image_id
```
Comment (jdkandersson) [inline src/charm.py:103]: this should probably be a snap - let's move to that and then we can have another review 🙂

File: .github/pull_request_template.md
Comment (yanksyoon) [inline .github/pull_request_template.md:1]: https://github.com/canonical/is-charms-template-repo To be inserted in the template.

File: .github/workflows/integration_test.yaml
Comment (yanksyoon) [inline .github/workflows/integration_test.yaml:1]: Fix in the template repository

File: .licenserc.yaml
Comment (yanksyoon) [inline .licenserc.yaml:1]: Include in repo

File: README.md
Comment (yanksyoon) [inline README.md:1]: Unfinished

File: charmcraft.yaml [hunk 12]
```diff
@@ -1,11 +1,75 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
 # This file configures Charmcraft.
-# See https://juju.is/docs/sdk/charmcraft-config for guidance.
+
+# This file configures Charmcraft.
```
Comment (yanksyoon) [inline charmcraft.yaml:5]: Duplicate line

File: charmcraft.yaml [hunk 13]
```diff
@@ -1,11 +1,75 @@
...
+title: Github runner image builder
+summary: Provide Github runners with images available to use.
+description: |
+  A [Juju](https://juju.is/) [charm](https://juju.is/docs/olm/charmed-operators) managing
+  runner images for
+  [self-hosted runners for GitHub Actions](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/about-self-hosted-runners).
+
+  A single unit of this charm provides pre-built images for Github runner operator to spawn and
+  register the self-hosted runners on Github. The charm will periodically rebuild and provide the
+  latest version of the image.
```
Comment (yanksyoon) [inline charmcraft.yaml:20]:
```suggestion
  latest version of the images.
```

File: charmcraft.yaml [hunk 14]
```diff
@@ -1,11 +1,75 @@
...
+  documentation: https://discourse.charmhub.io/t/github-runner-image-builder-documentation-overview
+  issues: https://github.com/canonical/github-runner-image-builder-operator/issues
+  source: https://github.com/canonical/github-runner-image-builder-operator
+  website: https://charmhub.io/github-runner-image-builder
 
 type: charm
 bases:
   - build-on:
-    - name: ubuntu
+    - name: "ubuntu"
```
Comment (yanksyoon) [inline charmcraft.yaml:31]: No need to be in quotes

File: charmcraft.yaml [hunk 15]
```diff
@@ -1,11 +1,75 @@
...
       channel: "22.04"
+      architectures:
+      - amd64
+      - arm64
+
+config:
+  options:
+    base-image:
+      type: string
+      default: jammy
```
Comment (yanksyoon) [inline charmcraft.yaml:47]: Default should be noble

File: charmcraft.yaml [hunk 16]
```diff
@@ -1,11 +1,75 @@
...
+      description: |
+        The base ubuntu OS image to use for the runners. Codename (e.g. "jammy") or version tag
+        (e.g. 22.04) is supported as input. Currently only supports LTS versions of jammy and
+        higher, i.e. jammy, noble.
```
Comment (yanksyoon) [inline charmcraft.yaml:51]: Example should prioritize noble

File: charmcraft.yaml [hunk 17]
```diff
@@ -1,11 +1,75 @@
...
+    build-interval:
+      type: int
+      default: 6
+      description: |
+        The interval in hours between each scheduled image builds. Setting this value to 0 means
+        image will not be rebuilt after the initial build.
+    experimental-openstack-clouds-yaml:
```
Comment (yanksyoon) [inline charmcraft.yaml:58]:
```suggestion
    openstack-clouds-yaml:
```

File: charmcraft.yaml [hunk 18]
```diff
@@ -1,11 +1,75 @@
...
+      type: string
+      default: ""
+      description: |
+        The clouds.yaml contents. See (https://docs.openstack.org/python-openstackclient/pike\
+          /configuration/index.html#clouds-yaml).
```
Comment (yanksyoon) [inline charmcraft.yaml:63]: Separate configuration for clouds yaml contents - model the configs rather than dump yamls.

File: charmcraft.yaml [hunk 19]
```diff
@@ -1,11 +1,75 @@
...
+    revision-history-limit:
+      type: int
+      default: 5
+      description: Number of image revisions to keep before deletion.
+
+actions:
+  build-image:
+    description: Builds a new image immediately.
```
Comment (yanksyoon) [inline charmcraft.yaml:71]: Delete

File: pyproject.toml [hunk 20]
```diff
@@ -12,31 +21,37 @@ log_cli_level = "INFO"
...
+max-line-length = 99
+max-doc-length = 99
 max-complexity = 10
+exclude = [".git", "__pycache__", ".tox", "build", "dist", "*.egg_info", "venv"]
+select = ["E", "W", "F", "C", "N", "R", "D", "H"]
+# Ignore W503, E501 because using black creates errors with this
+# Ignore D107 Missing docstring in __init__
+ignore = ["W503", "E501", "D107"]
+# D100, D101, D102, D103, D104, DCO020, DCO030: Ignore docstring style issues in tests
+per-file-ignores = ["tests/*:D100,D101,D102,D103,D104,D205,D212,DCO020,DCO030"]
```
Comment (yanksyoon) [inline pyproject.toml:41]: DCO020,DCO030 should be checked

File: requirements.txt [hunk 21]
```diff
@@ -1 +1,2 @@
 ops == 2.2.0
+openstacksdk>=3,<4
```
Comment (yanksyoon) [inline requirements.txt:2]: ops version is outdated.
Pin the openstacksdk version

File: src/charm.py [hunk 22]
```diff
@@ -1,103 +1,128 @@
...
 import logging
+from pathlib import Path
+from typing import Any
 
 import ops
 
-# Log messages can be retrieved using juju debug-log
+import builder
+import cron
+import image
```
Comment (yanksyoon) [inline src/charm.py:16]: image observer to be renamed

File: src/charm.py [hunk 23]
```diff
@@ -1,103 +1,128 @@
...
+
+        Args:
+            args: The CharmBase initialization arguments.
+        """
         super().__init__(*args)
-        self.framework.observe(self.on.httpbin_pebble_ready, self._on_httpbin_pebble_ready)
+        self.image_observer = image.Observer(self)
+        self.framework.observe(self.on.install, self._on_install)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.trigger, self._on_cron_trigger)
```
Comment (yanksyoon) [inline src/charm.py:42]: match the name of the event - event name should be on_cron_trigger

Comment (yanksyoon) [reply src/charm.py:42]: Rename to define what it triggers - the module can be renamed to be "task"

File: src/charm.py [hunk 24]
```diff
@@ -1,103 +1,128 @@
...
-        # Make Pebble reevaluate its plan, ensuring any services are started if enabled.
-        container.replan()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
-        self.unit.status = ops.ActiveStatus()
+        try:
+            return CharmState.from_charm(self)
+        except CharmConfigInvalidError as exc:
+            self.unit.status = ops.BlockedStatus(str(exc))
+            return None
```
Comment (yanksyoon) [inline src/charm.py:54]: Use decorator to catch the exception.

File: src/charm.py [hunk 25]
```diff
@@ -1,103 +1,128 @@
...
+        state = self._load_state()
+        if not state:
+            # Defer this event since on install should be re-triggered to setup dependencies for
+            # the charm. Since the charm goes into blocked state and the user reconfigures the
+            # state, config_changed event should be queued after the deferred on_install.
+            event.defer()
+            return
+
+        builder.configure_proxy(proxy=state.proxy_config)
+        builder.setup_builder()
```
Comment (yanksyoon) [inline src/charm.py:74]: rename to prepare - combine two funcs

File: src/charm.py [hunk 26]
```diff
@@ -1,103 +1,128 @@
...
+        self.unit.status = ops.MaintenanceStatus("Setting up Builder.")
+        state = self._load_state()
+        if not state:
+            # Defer this event since on install should be re-triggered to setup dependencies for
+            # the charm. Since the charm goes into blocked state and the user reconfigures the
+            # state, config_changed event should be queued after the deferred on_install.
+            event.defer()
+            return
+
+        builder.configure_proxy(proxy=state.proxy_config)
```
Comment (yanksyoon) [inline src/charm.py:73]: Why is this called here separately?

File: src/charm.py [hunk 27]
```diff
@@ -1,103 +1,128 @@
...
+
+        Args:
+            state: The charm state.
+
+        Returns:
+            The built image ID.
+        """
+        self.unit.status = ops.ActiveStatus("Building image.")
+        builder.configure_proxy(proxy=state.proxy_config)
+        output_path = Path("compressed.img")
```
Comment (yanksyoon) [inline src/charm.py:88]: Compressed.img should be a constant

File: src/charm.py [hunk 28]
```diff
@@ -1,103 +1,128 @@
...
-        self.unit.status = ops.ActiveStatus()
+        try:
+            return CharmState.from_charm(self)
+        except CharmConfigInvalidError as exc:
+            self.unit.status = ops.BlockedStatus(str(exc))
+            return None
 
-    def _on_config_changed(self, event: ops.ConfigChangedEvent):
-        """Handle changed configuration.
+    def _on_install(self, event: ops.InstallEvent) -> None:
```
Comment (yanksyoon) [inline src/charm.py:56]: Should setup cron here.

File: src/charm.py [hunk 29]
```diff
@@ -1,103 +1,128 @@
...
+        builder.run_builder(config=build_config)
+        with OpenstackManager(cloud_config=state.cloud_config) as manager:
+            upload_config = UploadImageConfig(
+                arch=state.image_config.arch,
+                app_name=self.app.name,
+                base=state.image_config.base_image,
+                num_revisions=state.revision_history_limit,
+                src_path=output_path,
+            )
+            image_id = manager.upload_image(config=upload_config)
```
Comment (yanksyoon) [inline src/charm.py:103]: Separate method.

File: src/charm.py [hunk 30]
```diff
@@ -1,103 +1,128 @@
...
+            self.image_observer.update_relation_data(image_id=image_id)
+        return image_id
```
Comment (yanksyoon) [inline src/charm.py:105]: Update_relation_data -> rename to business logic / property
No need return.

**Review by yanksyoon — COMMENTED** [review]
Initial comments during peer review.

File: src/utils.py
Comment (cbartz) [inline src/utils.py:1]: What about using a library such as https://github.com/jd/tenacity ?

File: tests/unit/test_state.py [hunk 31]
```diff
@@ -0,0 +1,453 @@
...
+
+    assert state._parse_build_interval(charm) == expected
+
+
+@pytest.mark.parametrize(
+    "revision_history, expected_err",
+    [
+        pytest.param("test", "An integer value for revision history is expected", id="non-int"),
+        pytest.param(
+            "-1", "Revision history must be greater than 1 and less than 100", id="netative"
```
Comment (cbartz) [inline tests/unit/test_state.py:265]:
```suggestion
            "-1", "Revision history must be greater than 1 and less than 100", id="negative"
```

File: tests/unit/test_utils.py
Comment (cbartz) [inline tests/unit/test_utils.py:1]: As mentioned above, using a library such as https://github.com/jd/tenacity would remove this

File: tests/unit/test_utils.py [hunk 32]
```diff
@@ -0,0 +1,78 @@
...
+
+import logging
+from unittest.mock import MagicMock
+
+import pytest
+
+from utils import retry
+
+
+def test_retry_with_logger():
```
Comment (cbartz) [inline tests/unit/test_utils.py:14]: Could you explain more about the need to test for logger=None and logger set.

File: tests/unit/test_utils.py [hunk 33]
```diff
@@ -0,0 +1,78 @@
...
+    """
+    arrange: given a function that raises exception decorated with retry.
```
Comment (cbartz) [inline tests/unit/test_utils.py:16]:
```suggestion
    arrange: given a function that raises exception decorated with retry using a local logger.
```

**Review by cbartz — COMMENTED** [review]
Great work. :rocket: Have some comments (mostly minor).

File: tests/unit/test_utils.py (hunk 33 above)
Comment (yanksyoon) [reply tests/unit/test_utils.py]: Fixed, thank you!

File: src-docs/openstack_manager.py.md
[the remark this answers is not in the document]
Comment (yanksyoon) [reply src-docs/openstack_manager.py.md:1]: These were artifacts remaining from last `src-docs`, since running src-docs doesn't automatically detect removed files. Thanks for the review!

File: README.md [hunk 34]
```diff
@@ -13,7 +13,7 @@ Use links instead.
 Charmhub package name: operator-template
 More information: https://charmhub.io/is-charms-template
 
-Describe your charm in one or two sentences.
+TODO: Describe your charm in one or two sentences.
```
[the remark this answers is not in the document]
Comment (yanksyoon) [reply README.md]: Updated, thank you!

File: src/image.py [hunk 35]
```diff
@@ -0,0 +1,81 @@
...
+            return CharmState.from_charm(self.charm)
+        except CharmConfigInvalidError as exc:
+            self.charm.unit.status = ops.BlockedStatus(str(exc))
+            return None
+
+    def _on_image_relation_joined(self, _: ops.RelationJoinedEvent) -> None:
+        """Handle the image relation joined event."""
+        state = self._load_state()
+        if not state:
+            return
```
[the remark this answers is not in the document]
Comment (yanksyoon) [reply src/image.py:61]: When the image gets built, it will propagate to existing relations hence it is okay to not defer this call :)

File: src/utils.py
Comment (yanksyoon) [reply src/utils.py:1]: Retry was no longer needed, thank you for the suggestion - i will resort to this next time :)

File: src/charm.py (hunk 22 above)
Comment (yanksyoon) [reply src/charm.py:15]: Would this still be the case if using `image.Observer`?

File: src/builder.py [hunk 36]
```diff
@@ -0,0 +1,348 @@
...
+            ],
+            timeout=5 * 60,
+            check=True,
+            user=UBUNTU_USER,
+        )
+    except (apt.PackageNotFoundError, subprocess.CalledProcessError) as exc:
+        raise DependencyInstallError from exc
+
+
+def _create_callback_script(config: CallbackConfig) -> None:
```
Comment (yanksyoon) [inline src/builder.py:134]: Should this be in `charm.py` since this is coupled with juju?

File: src/builder.py [hunk 37]
```diff
@@ -0,0 +1,348 @@
...
+    installed_interval = int(cron_args[1].split("/")[1])
+    installed_image_base = cron_args[13]
+    installed_cloud_name = cron_args[15]
+    installed_num_revisions = int(cron_args[17])
+    return (
+        installed_interval != interval
+        or installed_image_base != image_base.value
+        or installed_cloud_name != cloud_name
+        or num_revisions != installed_num_revisions
+    )
```
Comment (yanksyoon) [inline src/builder.py:266]: Cron class to define the parts? Or named tuples?

File: src/image.py [hunk 38]
```diff
@@ -0,0 +1,81 @@
...
+        """Load the charm state if valid, set charm to blocked if otherwise.
+
+        Returns:
+            Initialized charm state if valid charm state. None if invalid charm state was found.
+        """
+        try:
+            return CharmState.from_charm(self.charm)
+        except CharmConfigInvalidError as exc:
+            self.charm.unit.status = ops.BlockedStatus(str(exc))
+            return None
```
Comment (yanksyoon) [inline src/image.py:55]: Decorator here as well?

File: src/proxy.py [hunk 39]
```diff
@@ -0,0 +1,83 @@
...
+    if not proxy:
+        return
+    proxy_str = (proxy.http or proxy.https).replace("http://", "").replace("https://", "")
+    try:
+        subprocess.run(  # nosec: B603
+            ["/usr/bin/sudo", "snap", "set", "aproxy", f"proxy={proxy_str}"],
+            timeout=5 * 60,
+            check=True,
+            user=UBUNTU_USER,
+        )
```
Comment (yanksyoon) [inline src/proxy.py:81]: Does this snap need restart after configuration?

File: charmcraft.yaml [hunk 40]
```diff
@@ -1,11 +1,75 @@
...
+      architectures:
+      - amd64
+      - arm64
     run-on:
-    - name: ubuntu
+    - name: "ubuntu"
       channel: "22.04"
+      architectures:
+      - amd64
+      - arm64
```
[the remark this answers is not in the document]
Comment (yanksyoon) [reply charmcraft.yaml:40]: Yup, tested on the application side! :D

**Review by jdkandersson — APPROVED** [review]
It looks good - I couldn't see the comments from yesterday? Not sure if they are all addressed

**Comment by yanksyoon** [comment]
@jdkandersson somehow the comments are left as comments not as conversation 😭, i just have cloud config modelling left!

File: src/builder.py [hunk 41]
```diff
@@ -213,6 +213,10 @@ def run(config: state.BuilderRunConfig) -> None:
                     ">>",
                     str(OUTPUT_LOG_PATH),
                     "2>&1",
+                    "||",
+                    # Run the callback script without Openstack ID argument to let the charm know
+                    # about the error.
+                    str(config.callback_script),
```
Comment (jdkandersson) [inline src/builder.py:219]: Do you think this should be a different callback event rather than implicitly indicating failure by not passing a value?

File: src/charm.py [hunk 42]
```diff
@@ -1,103 +1,168 @@
...
+        env = " ".join(f'{key}="{val}"' for (key, val) in cur_env.items())
+        script_contents = f"""#! /bin/bash
+OPENSTACK_IMAGE_ID="$1"
+
+/usr/bin/juju-exec {self.unit.name} {env} {charm_dir}/dispatch
+"""
+        state.FAILED_CALLBACK_SCRIPT_PATH.write_text(script_contents, encoding="utf-8")
+        state.FAILED_CALLBACK_SCRIPT_PATH.chmod(0o755)
+
+    @charm_utils.block_if_invalid_config(defer=False)
```
Comment (amandahla) [inline src/charm.py:125]: Out of curiosity: Couldn't this be done similarly to this one?
https://github.com/canonical/synapse-operator/blob/main/src/charm_state.py#L64
So we could start having a pattern on how to block a charm in case of an invalid configuration.

File: src/image.py [hunk 43]
```diff
@@ -0,0 +1,67 @@
...
+
+        charm.framework.observe(
+            charm.on[IMAGE_RELATION].relation_joined, self._on_image_relation_joined
+        )
+
+    @charm_utils.block_if_invalid_config(defer=False)
+    def _on_image_relation_joined(self, _: ops.RelationJoinedEvent) -> None:
+        """Handle the image relation joined event."""
+        build_config = state.BuilderRunConfig.from_charm(charm=self.charm)
+        image_id = builder.get_latest_image(
```
Comment (amandahla) [inline src/image.py:50]: Is there any action that the user could take in case of GetLatestImageError?

File: src/charm.py (hunk 42 above)
Comment (yanksyoon) [reply src/charm.py:125]: Oh this is pretty cool, I've taken mine from Paas-app-charmer, https://github.com/canonical/paas-app-charmer/blob/main/paas_app_charmer/_gunicorn/charm_utils.py#L35.
I do like the way the way synapse-operator does to inject the state :D
But one thing though, is that I no longer use a big "charm state" but smaller dataclass objects (e.g. BuilderRunConfig) and load them at the time they are needed. Hence if I do inject those, I will need to inject multiple states which may not be desirable (I wouldn't want to couple the charm_util decorators to the state object).
I think the pattern is super helpful for charms with more generic single charm state instance, thank you for the suggestion!

File: src/image.py (hunk 43 above)
Comment (yanksyoon) [reply src/image.py:50]: I believe they would need to then investigate the error on why the application has raised an error - e.g. network error perhaps? (The application will return an empty string if no image was available / found :D )
Currently, there is no action that the user can take on the charm side to resolve this unfortunately, hence the fall-through to raise.
Thank you for the review!

File: src/builder.py [hunk 44]
```diff
@@ -0,0 +1,300 @@
...
+    "pipx",
+    # Required to build using pipx
+    "python3-dev",
+    "gcc",
+]
+CRON_PATH = Path("/etc/cron.d")
+CRON_BUILD_SCHEDULE_PATH = CRON_PATH / "build-runner-image"
+GITHUB_RUNNER_IMAGE_BUILDER = UBUNTU_HOME / ".local/bin/github-runner-image-builder"
+OPENSTACK_CLOUDS_YAML_PATH = UBUNTU_HOME / "clouds.yaml"
+OUTPUT_LOG_PATH = UBUNTU_HOME / "github-runner-image-builder.log"
```
[the remark this answers is not in the document]
Comment (yanksyoon) [reply src/builder.py:44]: Added to back-log and keeping it out of scope for now.

File: src/charm.py [hunk 45]
```diff
@@ -1,103 +1,160 @@
...
+
+        Args:
+            args: The CharmBase initialization arguments.
+        """
         super().__init__(*args)
-        self.framework.observe(self.on.httpbin_pebble_ready, self._on_httpbin_pebble_ready)
+        self.image_observer = image.Observer(self)
+        self.framework.observe(self.on.install, self._on_install)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.build_success, self._on_build_success)
```
[the remark this answers is not in the document]
Comment (jdkandersson) [reply src/charm.py:69]: Is this done as well?

[the remark this answers is not in the document]
Comment (yanksyoon) [reply src/charm.py:69]: No, working on it still!

**Review by jdkandersson — APPROVED** [review]
Is the upgrading the app after success and failure of builds done?

File: src/charm.py (hunk 45 above)
[the remark this answers is not in the document]
Comment (yanksyoon) [reply src/charm.py:69]: Now it is :D

File: charmcraft.yaml [hunk 46]
```diff
@@ -56,12 +60,48 @@ config:
       default: 6
       description: |
         The interval in hours between each scheduled image builds.
-    openstack-clouds-yaml:
+    os-auth-url:
```
Comment (jdkandersson) [inline charmcraft.yaml:63]: `os` is a bit confusing since it is so commonly refers to operating system, I would be ok with `openstack-*` it is a bit longer and quite clear

**Review by jdkandersson — DISMISSED** [review]

File: charmcraft.yaml (hunk 46 above)
Comment (yanksyoon) [reply charmcraft.yaml:63]: sounds good, thank you!

**Review by jdkandersson — DISMISSED** [review]

**Review by amandahla — APPROVED** [review]
LGTM, nice!

**Review by jdkandersson — DISMISSED** [review]

**Review by jdkandersson — DISMISSED** [review]

**Review by javierdelapuente — COMMENTED** [review]
Looks good. Just some nitpicking! (I would love to see this working!)

File: tests/integration/helpers.py [hunk 47]
```diff
@@ -0,0 +1,176 @@
...
+        time.sleep(10)
+    raise TimeoutError("No valid ssh connections found.")
+
+
+P = ParamSpec("P")
+R = TypeVar("R")
+S = Callable[P, R] | Callable[P, Awaitable[R]]
+
+
+async def wait_for(
```
[the remark this answers is not in the document]
Comment (yanksyoon) [reply tests/integration/helpers.py:124]: That's very neat! 
The difference for the custom `wait_for` function though, is that it can return the value after it has succeeded waiting :D

File: src/builder.py [hunk 48]
```diff
@@ -0,0 +1,290 @@
...
+        raise ImageBuilderInitializeError from exc
+
+
+def install_clouds_yaml(cloud_config: dict) -> None:
+    """Install clouds.yaml for Openstack used by the image builder.
+
+    Args:
+        cloud_config: The contents of clouds.yaml parsed as dict.
+    """
+    if not OPENSTACK_CLOUDS_YAML_PATH.exists():
```
[the remark this answers is not in the document]
Comment (yanksyoon) [reply src/builder.py:117]: That could also work :)

**Review by javierdelapuente — APPROVED** [review]

**Review by jdkandersson — APPROVED** [review]

**Review by javierdelapuente — APPROVED** [review]
